### PR TITLE
hpctoolkit: add version gpu and variant cuda

### PR DESCRIPTION
**Pull request by mwkrentel**
Add version for 'gpu' and variant for 'cuda'.  These are currently
available from the 'master-gpu' branch.

Add conflicts for old versions of gcc.

### CODE REVIEW

File: var/spack/repos/builtin/packages/hpctoolkit/package.py [hunk 1]
```diff
@@ -18,6 +18,7 @@ class Hpctoolkit(AutotoolsPackage):
     git      = "https://github.com/HPCToolkit/hpctoolkit.git"
 
     version('develop', branch='master')
+    version('gpu', branch='master-gpu')
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/hpctoolkit/package.py]: Note that we support `master` as a special, ahead-of-everything-except-develop version now.  `gpu` is not special and will compare as earlier than your other yyyy.mm.dd versions.  You could call these versions `master` and `master-gpu`, which might be more understandable to HPCToolkit users.

Comment (mwkrentel) [reply var/spack/repos/builtin/packages/hpctoolkit/package.py]: I didn't realize that `master` had been added as a special name (along
with head and trunk, I see).  I'll change develop to master.

As for `gpu`, I'll check with @jmellorcrummey.  But I'm inclined to
keep it as gpu.  The name of the branch has already changed a couple
times.  `master-gpu` just means the merge of master and gpu, before
it's ready to be merged into master (at which time, we can drop the gpu
branch).

**Comment by jmellorcrummey** [comment]
The gpu branch is a preview. We prefer the simple name to master-gpu. We want the name in spack to facilitate widely deploying the preview. Over the next few months, we will remove this version when we merge the branch with master.

**Comment by mwkrentel** [comment]
@tgamblin Yes, I do understand that `2018.11.05 > gpu`.  I can
probably deal with any ordering as long as it's stable and
deterministic.  The `cuda` variant only works with the `gpu` branch.
I wanted to write a conflict as:

```
conflicts('+cuda', when='not @gpu')
```

But that's not allowed, at least not yet.  This is one reason I've
asked about boolean ops in `when` clauses and specs.

Instead, I used the ordering to write the conflict as this, which relies on
`develop > yyyy.mm.dd > gpu`.

```
conflicts('+cuda', when='@2018.0.0:')
```

**Comment by tgamblin** [comment]
Ok, sounds reasonable -- thanks!

I understand the issue with booleans in specs.  I am not sure when we can get to that -- but there will be a survey of ST teams early next year where we'll try to see what ST folks want as the highest priorities.  Really, for this problem, we should add `when` to variants rather than using conflicts, but that's another feature we'll need to get on the roadmap.

**Pull request merged**